Rex superswap 127

gitChimp88 opened the pull request:
(work in progress)
Port uniswap contracts so we can use solidity 0.8
Create RexSuperSwap contract
Create test for contract

gitChimp88 committed:
Create RexSuperSwap.sol

Bring RexSuperSwap contract in to this repo to begin with

gitChimp88 committed:
Add deployment test for RexSuperSwap

Add test for RexSuperSwap to test deployment, Install uniswap sdk - There is an error for incompatible solidity versions between dependency and RexSuperSwap.

gitChimp88 committed:
Port uniswap contracts and update them to sol 0.8

Port uniswap contracts that we need to use for super swap, add some more test boilerplate

gitChimp88 committed:
Test swap function from REXSuperSwap

Test passing for swap functionality, need to test balances and different tokens.

gitChimp88 committed:
Add check for account balances in test

Add check for account balances in test
Add call to coingecko api to get minimum amount out of swap
Add event to indicate swap just made

gitChimp88 committed:
Fork from latest block number

fork from latest block number, change swap to be between matic and usd

gitChimp88 committed:
Try different transfer method, add console logs to track balance of contract

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,108 @@
+    // Approve the router to spend token supplied (fromBase).
+    TransferHelper.safeApprove(fromBase, address(swapRouter), amountIn);
+
+    console.log("balance of Maticx before swap - ", _from.balanceOf(address(this)));
```
This will come out to be 0 because you downgraded MATICx to native MATIC on line 59, to get the balance of the matic you have, do:
```
console.log("balance of MATIC before swap - ", address(this).balance);

gitChimp88 replied:
thanks for the tip, now changed

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,108 @@
+    // Execute the swap
+    amountOut = swapRouter.exactInput(params);
+
+    console.log("balance of Maticx after swap - ", _from.balanceOf(address(this)));
```
Again, after the swap, you won't have MATICx, you're going to have native MATIC, check balance with `address(this).balance` 

(actually, not sure how uniswap v3 behaves, it might give you wrapped matic in which case youll check the balance as:
```
ERC20(_from.getUnderlyingToken()).balanceOf(address(this))
```

gitChimp88 replied:
changed

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,108 @@
+
+    console.log("balance of Maticx after swap - ", _from.balanceOf(address(this)));
+
+    emit SwapJustMade(amountOut);
```
I don't think this is needed since Uniswap v3 emits events, remove this event

gitChimp88 replied:
Now removed

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,108 @@
+
+    // Step 1: Get underlying tokens and verify path
+    address fromBase = _from.getUnderlyingToken();
+    address toBase = _to.getUnderlyingToken();
```
You're missing code here to handle native supertokens, see this code: https://github.com/Ricochet-Exchange/ricochet-protocol/blob/v2/contracts/REXTwoWayMarket.sol#L396

gitChimp88 replied:
Okay i've now added this check

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,108 @@
+    console.log("starting balance of Maticx - ", _from.balanceOf(address(this)));
+
+    // Step 3: Downgrade
+    _from.downgrade(amountIn);
```
Same here, you're not handling the native supertoken case where the downgrades not needed, swapping RIC to USDCx for example, RIC doens't need to be downgraded, see the code I link to in the comment above

gitChimp88 replied:
Now handling correctly

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,108 @@
+
+    // Step 5: Upgrade and send tokens back
+    TransferHelper.safeApprove(address(toBase), address(_to), amountOut);
+    _to.upgrade(amountOut);
```
same here with the upgrade, you have to account for native supertokens, like if they're swapping USDCx to RIC, no upgrade required on RIC, see this code snippet: https://github.com/Ricochet-Exchange/ricochet-protocol/blob/v2/contracts/REXTwoWayMarket.sol#L443,

mikeghen replied:
You also need to check here if the `_to` token is MATICx, if so, you have to use `upgradeByETH` method, see the code snippet here: https://github.com/Ricochet-Exchange/ricochet-protocol/blob/v2/contracts/REXTwoWayMarket.sol#L434

gitChimp88 replied:
Great thanks, now handling correctly and added the check for MATICX

mikeghen reviewed (changes requested):
see comments, I don't think this is set up right to handle MATICx, which downgrades to _native_ MATIC and needs to be upgraded with `upgradeByETH`, also this doesn't handle native supertokens, see my comments

gitChimp88 committed:
Add logic to handle native tokens

Logic added for native tokens downgrade and upgrade, also upgrade of maticx

gitChimp88 committed:
fix upgrade and tokens are now being transferred

gitChimp88 committed:
Add approve function to super swap

add approve function for unlimited approvals rather than decimal conversions, has underlying to and from variables to deal with native tokens, add test for native token

gitChimp88 committed:
Add test for swapping ethx to usdcx

gitChimp88 committed:
Add try catch to swap function

Add try catch and emit events on swap errors, run tests from a specific block

gitChimp88 committed:
Test native ric swap and swap from matic - wbtc

shreyaspapi commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,147 @@
+  using SafeERC20 for ERC20;
+  ISwapRouter02 public immutable swapRouter;
+
+  address public constant MATICX = 0x3aD736904E9e65189c3000c7DD2c8AC8bB7cD4e3;
```
Can we add this in the constructor so we don't have issues deploying to other network.

shreyaspapi replied:
Maybe something like a `superNativeToken` var, a while back I was working on it here - https://github.com/Ricochet-Exchange/ricochet-super-swap/pull/2

gitChimp88 replied:
Alright good catch, I added the matic address to the contract constructor as you suggested. Can you think of any other tests we should run on this swap or anything that could be missing?

shreyaspapi commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,147 @@
+
+import { ISuperToken } from "@superfluid-finance/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol";
+import "./TransferHelper.sol";
+import "./interfaces/ISwapRouter02.sol";
```
We need a note here for future reference, 

```
- You will manually have to change `@uniswap/swap-router-contracts/contracts/interfaces/ISwapRouter02.sol` `^0.7.6` ==> `^0.8.0`
```

shreyaspapi commented:
@gitChimp88 I pushed the local changes I had for https://github.com/Ricochet-Exchange/ricochet-super-swap/pull/2

gitChimp88 committed:
Add matic address to the contract constructor

shreyaspapi commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,143 @@
+  event ErrorOnSwap(string reason);
+  event ReturnDataEvent(bytes returnData);
+
+  constructor(ISwapRouter02 _swapRouter, address _maticAddress) {
```
`_maticAddress` -> `superNativeToken`, just to avoid confusion with other chains.

gitChimp88 replied:
Ah yeah I understand now, variable has been changed!

shreyaspapi commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,143 @@
+contract RexSuperSwap {
+  using SafeERC20 for ERC20;
+  ISwapRouter02 public immutable swapRouter;
+  address public maticAddress;
```
Same, change variable name.

shreyaspapi reviewed (changes requested):
Cool this looks great! Just one lil change + Can you also add a test for `RIC -> MaticX`

gitChimp88 committed:
Change variable name and add extra test

gitChimp88 commented:
> Cool this looks great! Just one lil change + Can you also add a test for `RIC -> MaticX`

I added the test but it fails at the moment because there's no liquidity on uniswap RIC -> Matic

shreyaspapi commented on hardhat.config.ts:
```diff
@@ -69,7 +69,7 @@ const config: HardhatUserConfig = {
         accounts: process.env.POLYGON_PRIVATE_KEY !== undefined ? [process.env.POLYGON_PRIVATE_KEY] : [],
         enabled: true,
-        blockNumber: 22877930     // Essential for mainnet forking !!
+        // blockNumber: 22877930     // Essential for mainnet forking !!
```
```suggestion
        blockNumber: 22877930     // Essential for mainnet forking !!
```

gitChimp88 replied:
Alright unwanted comments removed and it's now forking from this block number

shreyaspapi reviewed (approved):
Tests looks good to me, we need to test RIC -> MATICx @mikeghen, Also can you remove the unwanted comments in the test file @gitChimp88 .

gitChimp88 committed:
Remove unwanted comments in test file

mikeghen commented:
@gitChimp88 I just made some liquidity for RIC-MATIC on uniswap

gitChimp88 committed:
Test user can swap small amounts and large amounts

gitChimp88 committed:
add deploy script for superswap

shreyaspapi commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,142 @@
+
+    // Upgrade if it's not a native SuperToken
+    if (_hasUnderlyingTo) {
+      if (address(_to) == superNativeToken) {
```
@gitChimp88 We don't need this condition, here I wrongly assumed that Super Matic has underlaying token as 0x (native token). As a result the Matic <> Ric test might be failing.

gitChimp88 committed:
Delete unnecessary check for matic token

Delete unneccessary if statement and different upgrade method as matic is downgraded to wrapped matic, specified specific block (which has the required liquidity for RIC)

mikeghen commented on contracts/superswap/RexSuperSwap.sol:
```diff
@@ -0,0 +1,136 @@
+      emit ErrorOnSwap(reason);
+    } catch (bytes memory returnData) {
+      emit ReturnDataEvent(returnData);
+    }
```
There's a problem here, if the swap reverts the contract will eat the input tokens and not return them to the user, I would just not use a try catch here at all, if swapRouter.exactInput(params) reverts, revert the whole thing

gitChimp88 committed:
Add withdraw function to super swap

Withdraw function to convert wmatic to matic so we can upgrade to super token

shreyaspapi commented:
Closes #127

mikeghen commented:
@gitChimp88 @shreyaspapi, hows this going? I took a look at the last commit, it looks like a test still needs to done. Lmk how I can help you.
